File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/features/MonsterFeature.java
package com.joeshuff.dddungeongenerator.generator.features;

import com.joeshuff.dddungeongenerator.Logs;
import com.joeshuff.dddungeongenerator.generator.dungeon.Dungeon;
import com.joeshuff.dddungeongenerator.generator.dungeon.Modifier;
import com.joeshuff.dddungeongenerator.generator.monsters.Bestiary;
import com.joeshuff.dddungeongenerator.generator.monsters.Monster;

import java.util.Random;

public class MonsterFeature implements RoomFeature {

    int size;
    Monster selectedMonster;

    private static transient double DEFAULT_BOSS_CHANCE = 0.3d;

    transient Modifier modifier;
    transient Random rnd;
    boolean isBoss = false;

    public MonsterFeature(String seed, Modifier m) {
        this.modifier = m;
        this.rnd = new Random(Dungeon.stringToSeed(seed));
        size = Math.max(1, (int) ((rnd.nextGaussian() * 2) + m.getMonsterAverageGroupSize()));
        generateMonster();
    }

    private void generateMonster() {
        if (size == 1) {
            if (rnd.nextDouble() <= (DEFAULT_BOSS_CHANCE * modifier.getBossChanceModifier())) {
                selectedMonster = Bestiary.getBoss(rnd, modifier, false);

                if (selectedMonster != null) {
                    isBoss = true;
                } else {
                    Logs.i("MonsterFeature", "EMPTY BOSS MONSTER", null);
                }
            }
        }

        if (selectedMonster != null) return;

        selectedMonster = Bestiary.getMonster(rnd, modifier, false);

        if (selectedMonster == null) {
            Logs.i("MonsterFeature", "EMPTY MONSTER", null);
        }
    }

    @Override
    public String getFeatureDescription() {
        if (isBoss) {
            return "there is a boss in this room. A " + selectedMonster.getName() + " resides here";
        }

        if (size > 1) {
            return "there are " + size + " " + selectedMonster.getName() + "'s in this room, be careful!";
        } else {
            return "there is a " + selectedMonster.getName() + " in this room. Be careful!";
        }
    }

    @Override
    public String getFeatureName() {
        return "Monster(s)";
    }

    @Override
    public int getPageForMoreInfo() {
        return 0;
    }

    public Monster getSelectedMonster() {
        return selectedMonster;
    }

    public int getSize() {
        return size;
    }

    public boolean isBoss() {
        return isBoss;
    }
}

File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/features/MonsterFeature.kt
package com.joeshuff.dddungeongenerator.generator.features

import com.joeshuff.dddungeongenerator.generator.dungeon.Dungeon
import com.joeshuff.dddungeongenerator.generator.dungeon.Modifier
import com.joeshuff.dddungeongenerator.generator.monsters.Bestiary
import com.joeshuff.dddungeongenerator.generator.monsters.Monster
import com.joeshuff.dddungeongenerator.util.Logs
import java.util.*
import kotlin.math.max

class MonsterFeature(seed: String?, @field:Transient var modifier: Modifier) : RoomFeature {
    var size: Int
    var selectedMonster: Monster? = null

    @Transient
    var rnd: Random = Random(Dungeon.stringToSeed(seed))

    companion object {
        @Transient
        private val DEFAULT_BOSS_CHANCE = 0.3
    }

    init {
        size = max(1, (rnd.nextGaussian() * 2 + modifier.monsterAverageGroupSize).toInt())
        generateMonster()
    }

    var isBoss = false
    private fun generateMonster() {
        if (size == 1) {
            if (rnd.nextDouble() <= (DEFAULT_BOSS_CHANCE * modifier.bossChanceModifier)) {
                selectedMonster = Bestiary.getBoss(rnd, modifier, false)
                if (selectedMonster != null) {
                    isBoss = true
                } else {
                    Logs.i("MonsterFeature", "EMPTY BOSS MONSTER", null)
                }
            }
        }

        if (selectedMonster != null) return
        selectedMonster = Bestiary.getMonster(rnd, modifier, false)

        if (selectedMonster == null) {
            Logs.i("MonsterFeature", "EMPTY MONSTER", null)
        }
    }

    override fun getFeatureDescription(): String {
        if (isBoss) {
            return "there is a boss in this room. A " + selectedMonster!!.name + " resides here"
        }
        
        return if (size > 1) {
            "there are " + size + " " + selectedMonster!!.name + "'s in this room, be careful!"
        } else {
            "there is a " + selectedMonster!!.name + " in this room. Be careful!"
        }
    }

    override fun getFeatureName(): String {
        return "Monster(s)"
    }

    override fun getPageForMoreInfo(): Int {
        return 0
    }
}